src/main.rs

```diff
@@ -0,0 +1,289 @@
+    let state = state.read().map_err(|_| Error::LockFailure)?;
+    let current_epoch = state.epoch;
+    // FIXME: return the end of the current epoch
+    let next_epoch_time = "unknown".to_owned();
+    let max_points = MAX_POINTS;
+    let public_key = state.server.get_public_key()
+        .serialize_to_bincode()
+        .map_err(Error::Oprf)?;
+    let public_key = BASE64.encode(public_key);
+    let response = InfoResponse {
+        current_epoch,
+        next_epoch_time,
+        max_points,
+        public_key,
+    };
+    debug!("send: {response:?}");
+    Ok(Json(response))
+}
+
+/// Advance to the next epoch on a timer
+async fn epoch_update_loop(state: OPRFState, epochs: EpochRange) {
+    let interval = std::time::Duration::from_secs(5);
+    info!("rotating epoch every {} seconds", interval.as_secs());
+
+    loop {
+        // Wait until the current epoch ends.
+        tokio::time::sleep(interval).await;
+
+        // Acquire exclusive access to the oprf state.
+        // Panics if this fails, since processing requests with an
+        // expired epoch weakens user privacy.
+        let mut s = state.write().expect("Failed to lock OPRFState");
+
+        // Puncture the current epoch so it can no longer be used.
+        let old_epoch = s.epoch;
+        s.server.puncture(old_epoch).expect("Failed to puncture current epoch");
+
+        // Advance to the next epoch.
+        let new_epoch = old_epoch + 1;
+        if epochs.contains(&new_epoch) {
+            // Server is already initialized for this one.
+            s.epoch = new_epoch;
+        } else {
+            info!("Epochs exhausted! Rotating OPRF key");
+            // Panics if this fails. Puncture should mean we can't
+            // violate privacy through further evaluations, but we
+            // still want to drop the inner state with its private key.
+            *s = OPRFServer::new(&epochs)
+                .expect("Could not initialize new PPOPRF state");
+        }
+        info!("epoch now {}", s.epoch);
+    }
+}
+
+/// Initialize an axum::Router for our web service
+///
+/// Having this as a separate function makes testing easier.
+fn app(oprf_state: OPRFState) -> Router {
+    Router::new()
+        // Friendly default route to identify the site
+        .route("/", get(|| async { "STAR randomness server\n" }))
+        // Main endpoint
+        .route("/randomness", post(randomness))
+        .route("/info", get(info))
+        .with_state(oprf_state)
+        // Logging must come after active routes
+        .layer(tower_http::trace::TraceLayer::new_for_http())
+}
+
+#[tokio::main]
+async fn main() {
+    // Start logging
+    // The default subscriber respects filter directives like `RUST_LOG=info`
+    //let filter = tracing_subscriber::EnvFilter::from_default_env();
+    //let logger = tracing_subscriber::FmtSubscriber::new();
+    //tracing::subscriber::set_global_default(logger).unwrap();
+    tracing_subscriber::fmt::init();
```

Comment:
Do we use `tracing` in our dependencies (i.e. `sta-rs` or `constellation`)? It could be more straightforward to just use the `log` crate if we don't, or if we don't care to see traces/logs from dependencies

Reply:
Scratch that, I see that you're using the tower trace middleware.